clients/drcachesim/tests/cache_replacement_policy_unit_test.cpp

```diff
@@ -105,9 +124,107 @@ unit_test_cache_lru_four_way()
     cache_lru_test.access_and_check_lru(ADDRESS_E, 2); // A E c D
 }
 
+void
+unit_test_cache_fifo_four_way()
+{
+    cache_fifo_test_t cache_fifo_test;
+    cache_fifo_test.initialize_cache(/*associativity=*/4, /*line_size=*/32,
+                                     /*total_size=*/256);
+    const addr_t ADDRESS_A = 0;
+    const addr_t ADDRESS_B = 64;
+    const addr_t ADDRESS_C = 128;
+    const addr_t ADDRESS_D = 192;
+    const addr_t ADDRESS_E = 72;
+    const addr_t ADDRESS_F = 130;
+
+    assert(cache_fifo_test.get_block_index(ADDRESS_A) ==
+           cache_fifo_test.get_block_index(ADDRESS_B));
+    assert(cache_fifo_test.get_block_index(ADDRESS_B) ==
+           cache_fifo_test.get_block_index(ADDRESS_C));
+    assert(cache_fifo_test.get_block_index(ADDRESS_C) ==
+           cache_fifo_test.get_block_index(ADDRESS_D));
+    assert(cache_fifo_test.get_block_index(ADDRESS_D) ==
+           cache_fifo_test.get_block_index(ADDRESS_E));
+    assert(cache_fifo_test.get_block_index(ADDRESS_E) ==
+           cache_fifo_test.get_block_index(ADDRESS_F));
+
+    // Lower-case letter shows the way that is to be replaced after the access.
+    cache_fifo_test.access_and_check_fifo(ADDRESS_A, 1); // A x X X
+    cache_fifo_test.access_and_check_fifo(ADDRESS_B, 2); // A B x X
+    cache_fifo_test.access_and_check_fifo(ADDRESS_C, 3); // A B C x
+    cache_fifo_test.access_and_check_fifo(ADDRESS_D, 0); // a B C D
+    cache_fifo_test.access_and_check_fifo(ADDRESS_A, 0); // a B C D
+    cache_fifo_test.access_and_check_fifo(ADDRESS_A, 0); // a B C D
+    cache_fifo_test.access_and_check_fifo(ADDRESS_A, 0); // a B C D
+    cache_fifo_test.access_and_check_fifo(ADDRESS_E, 0); // e B C D
```

Comment:
Bug: this is not FIFO behavior: the next to replace is way number 1, where B is.  The fact that this passed means something is wrong with the test: it looks like it's because there is tag overlap, so this is not actually a miss.  The E address of 72 is the same tag as the address of B!  So this is a hit; it's just like accessing B.

Please add checks that the addresses all have distinct tags.

Reply:
updated the test sequence with addresses that have distinct tags.